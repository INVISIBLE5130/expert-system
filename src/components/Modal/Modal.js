import React, {useEffect, useRef, useState} from 'react'
import {
    ModalWrapper,
    ModalWrapperWindow,
    ModalWrapperWindowBody,
    ModalWrapperWindowBodyAnswerButton,
    ModalWrapperWindowBodyAnswerButtons,
    ModalWrapperWindowBodyAnswerButtonsItem,
    ModalWrapperWindowBodyAnswerInput,
    ModalWrapperWindowBodyQuestion,
    ModalWrapperWindowHeader,
    ModalWrapperWindowHeaderClose,
    ModalWrapperWindowHeaderTitle, ModalWrapperWindowResultTitle
} from "./Modal.styled";
import Close from '../../assets/close.svg'
import {remoteData as rData} from "../../SES_KB";

function Modal(props) {
    const {
        parsedCode,
        setConsultingStatus,
    } = props
    const modalRef = useRef(null)
    const modalWindowRef = useRef(null)
    const [remoteData, setRemoteData] = useState([])
    const [searchKey, setSearchKey] = useState('')
    const [data, setData] = useState({
        searchKey: '',
        localKeys: {}
    })

    useEffect(() => {
        if (parsedCode !== {}) {
            setRemoteData(Object.entries(parsedCode?.data))
        }
    }, [parsedCode])

    const searchKeyHandler = e => {
        const value = e.target.value
        setSearchKey(value)
    }

    const setSearchKeyHandler = () => {
        setData({
            searchKey: searchKey,
            localKeys: {}
        })
    }

    const closeModalHandler = () => {
        setConsultingStatus(false)
        setSearchKey('')
        setData({
            searchKey: '',
            localKeys: {}
        })
    }

    const setQuestion = (question, buttons, q) => {
        return (
            <>
                <ModalWrapperWindowBodyQuestion>
                    {question}
                </ModalWrapperWindowBodyQuestion>
                <ModalWrapperWindowBodyAnswerButtons>
                    {
                        buttons?.map(button => {
                            return (
                                <ModalWrapperWindowBodyAnswerButtonsItem
                                    onClick={() => answerIndexHandler({
                                        question: q,
                                        answer: button
                                    })}
                                >
                                    {button}
                                </ModalWrapperWindowBodyAnswerButtonsItem>
                            )
                        })
                    }
                </ModalWrapperWindowBodyAnswerButtons>
            </>
        )
    }

    const tryAgainHandler = () => {
        setSearchKey('')
        setData({
            searchKey: '',
            localKeys: {}
        })
        setRemoteData(Object.entries(rData?.data))
    }

    const showResult = result => {
        return (
            <>
                <ModalWrapperWindowResultTitle>
                    {'Expertise result:'}
                </ModalWrapperWindowResultTitle>
                <ModalWrapperWindowResultTitle>
                    {result}
                </ModalWrapperWindowResultTitle>
                <ModalWrapperWindowBodyAnswerButtons>
                    <ModalWrapperWindowBodyAnswerButtonsItem
                        onClick={tryAgainHandler}
                    >
                        {'Try again'}
                    </ModalWrapperWindowBodyAnswerButtonsItem>
                    <ModalWrapperWindowBodyAnswerButtonsItem
                        onClick={closeModalHandler}
                    >
                        {'Exit'}
                    </ModalWrapperWindowBodyAnswerButtonsItem>
                </ModalWrapperWindowBodyAnswerButtons>
            </>
        )
    }

    const answerIndexHandler = (d) => {
        setRemoteData(remoteData.filter(e => e[0] !== d.question))
        console.log(d)
        setData({
            searchKey: searchKey,
            localKeys: {
                ...data.localKeys,
                [d.question]: d.answer
            }
        })
    }

    console.log(data)

    useEffect(() => {
        function handleClickOutside(event) {
            const target = event.target
            if (
                modalRef?.current?.contains(target) &&
                !modalWindowRef?.current?.contains(target)
            ) {
                setConsultingStatus(false)
                setSearchKey('')
                setData({
                    searchKey: '',
                    localKeys: {}
                })
            }
        }

        document.addEventListener("mousedown", handleClickOutside);
        return () => {
            document.removeEventListener("mousedown", handleClickOutside);
        };
    }, [modalRef, modalWindowRef]);

    console.log(remoteData)

    return (
        <ModalWrapper
            ref={modalRef}
        >
            <ModalWrapperWindow
                ref={modalWindowRef}
            >
                <ModalWrapperWindowHeader>
                    <ModalWrapperWindowHeaderTitle>
                        {'Consulting'}
                    </ModalWrapperWindowHeaderTitle>
                    <ModalWrapperWindowHeaderClose
                        src={Close}
                        onClick={closeModalHandler}
                    />
                </ModalWrapperWindowHeader>
                <ModalWrapperWindowBody>
                    {
                        data.searchKey === ''
                            ? <>
                                <ModalWrapperWindowBodyQuestion>
                                    {'What is your search key?'}
                                </ModalWrapperWindowBodyQuestion>
                                <ModalWrapperWindowBodyAnswerInput
                                    value={searchKey}
                                    onChange={searchKeyHandler}
                                />
                                <ModalWrapperWindowBodyAnswerButton
                                    onClick={setSearchKeyHandler}
                                >
                                    {'Send'}
                                </ModalWrapperWindowBodyAnswerButton>
                            </>
                            : remoteData.length !== 0
                            ? setQuestion(remoteData[0][1]?.question, remoteData[0][1]?.allows, remoteData[0][0])
                            : showResult(JSON.stringify(parsedCode.rules.filter(e => JSON.stringify(e.conditions) === JSON.stringify(data.localKeys))[0]?.action))
                    }
                </ModalWrapperWindowBody>
            </ModalWrapperWindow>
        </ModalWrapper>
    )
}

export default Modal