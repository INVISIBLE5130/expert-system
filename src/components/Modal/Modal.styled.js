import styled from 'styled-components'
import {color} from "../../config";
import {HeaderWrapperUploadButton} from "../Header/Header.styled";

export const ModalWrapper = styled.div`
  top: 0;
  left: 0;
  z-index: 1;
  width: 100%;
  display: flex;
  position: fixed;
  min-height: 100vh;
  align-items: center;
  justify-content: center;
  background-color: rgba(0,0,0,0.25);
`

export const ModalWrapperWindow = styled.div`
  width: 90%;
  height: 80vh;
  padding: 15px;
  display: flex;
  flex-direction: column;
  background-color: white;
`

export const ModalWrapperWindowHeader = styled.div`
  width: 100%;
  display: flex;
  align-items: center; 
  justify-content: center;
`

export const ModalWrapperWindowHeaderTitle = styled.p`
  font-size: 24px;
  font-weight: 500;
  font-family: Arial, sans-serif;
`

export const ModalWrapperWindowHeaderClose = styled.img`
  width: 15px;
  height: 15px;
  cursor: pointer;
  margin: 0 0 0 auto;
`

export const ModalWrapperWindowBody = styled.div`
  height: 100%;
  display: flex;
  overflow-y: auto;
  align-items: center;
  flex-direction: column;
  justify-content: center;
`

export const ModalWrapperWindowBodyQuestion = styled.p`
  font-size: 18px;
  text-align: center;
  margin-bottom: 25px;
  font-family: Arial, sans-serif;
`

export const ModalWrapperWindowBodyAnswerInput = styled.input`
  width: 50%;
  padding: 5px;
  outline: none;
  align-self: center;
  border: 1px solid ${color.border};
`

export const ModalWrapperWindowBodyAnswerButtons = styled.div`
  display: flex;
`

export const ModalWrapperWindowBodyAnswerButtonsItem = styled(HeaderWrapperUploadButton)`
  margin-left: 10px;
  
  &:first-child {
    margin-left: unset;
  }
`

export const ModalWrapperWindowBodyAnswerButton = styled(HeaderWrapperUploadButton)`
  margin-top: 25px;
  align-self: center;
`

export const ModalWrapperWindowResultTitle = styled(ModalWrapperWindowHeaderTitle)`
  margin-bottom: 15px;
`