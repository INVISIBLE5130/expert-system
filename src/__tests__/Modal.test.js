import React from "react";
import renderer from 'react-test-renderer'
import Modal from '../components/Modal/Modal'

it("Check modal", () => {
    const ModalComponent = renderer.create(<Modal/>).toJSON();
    expect(ModalComponent).toMatchSnapshot();
});