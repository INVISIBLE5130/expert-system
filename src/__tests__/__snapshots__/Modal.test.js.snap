// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check modal 1`] = `
<div
  className="sc-fodVek gzdfup"
>
  <div
    className="sc-fFucqa huRzSq"
  >
    <div
      className="sc-bkzYnD crtOkE"
    >
      <p
        className="sc-idOiZg eJCtKS"
      >
        Consulting
      </p>
      <img
        className="sc-dIUeWJ gbzGpW"
        onClick={[Function]}
        src="close.svg"
      />
    </div>
    <div
      className="sc-hHfuMS kBfJxG"
    >
      <p
        className="sc-dmlqKv fRTqKP"
      >
        What is your search key?
      </p>
      <input
        className="sc-kfzBvY FbVbX"
        onChange={[Function]}
        value=""
      />
      <button
        className="sc-crrszt sc-iwyWTf jVhSAi eJjZwV"
        onClick={[Function]}
      >
        Send
      </button>
    </div>
  </div>
</div>
`;
